backend/service/chaos/experimentation/terminator/terminator.go

```diff
@@ -0,0 +1,176 @@
+		// fairness, as checking the termination conditions for one experiment should not be delaying the checks for another experiment
+		// unless we're under very heavy load.
+		go func(configType string) {
+			trackedExperiments := map[uint64]context.CancelFunc{}
+			ticker := time.NewTicker(m.outerLoopInterval)
+
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				case <-ticker.C:
+					es, err := m.store.GetExperiments(context.Background(), configType, experimentationv1.GetExperimentsRequest_STATUS_RUNNING)
+					if err != nil {
+						m.log.Errorw("failed to retrieve experiments from experiment store", "err", err, "enableConfigTypes", m.enabledConfigTypes)
+						continue
+					}
+
+					activeExperiments := m.monitorNewExperiments(es, trackedExperiments)
+
+					// For all experiments that we're tracking that no longer appear to be active, cancel the goroutine
+					// and clean it up.
+					for e, cancel := range trackedExperiments {
+						if _, ok := activeExperiments[e]; !ok {
+							cancel()
+							delete(trackedExperiments, e)
+						}
+					}
+				}
+			}
+		}(configType)
+	}
+}
+
+// monitorNewExperiments iterates over all the provided experiments, spawning a goroutine to montior each experiment that
+// doesn't already have a monitoring routine. Returns a set containing all the active experiment ids for further processing.
+func (m *monitor) monitorNewExperiments(es []*experimentationv1.Experiment, trackedExperiments map[uint64]context.CancelFunc) map[uint64]struct{} {
+	// For each active experiment, create a monitoring goroutine if necessary.
+	activeExperiments := map[uint64]struct{}{}
+	for _, e := range es {
+		activeExperiments[e.Id] = struct{}{}
+		if _, ok := trackedExperiments[e.Id]; !ok {
+			ctx, cancel := context.WithCancel(context.Background())
+			trackedExperiments[e.Id] = cancel
+
+			m.activeMonitoringRoutines.inc()
+			go func() {
+				defer m.activeMonitoringRoutines.dec()
+				m.monitorSingleExperiment(ctx, e)
+			}()
+		}
+	}
+	return activeExperiments
+}
+
+func (m *monitor) monitorSingleExperiment(ctx context.Context, e *experimentationv1.Experiment) {
+	ticker := time.NewTicker(m.perExperimentCheckInterval)
+	terminated := false
+
+	dany := &ptypes.DynamicAny{}
+	err := ptypes.UnmarshalAny(e.Config, dany)
+	if err != nil {
+		m.log.Errorw("failed to unmarshal experiment", "id", e.Id)
+		m.marshallingErrors.Inc(1)
+		return
+	}
+
+	for {
+		select {
+		case <-ticker.C:
+			if terminated {
+				// Note that we don't exit the goroutine here and instead wait for the outer
+				// monitoring loop to detect that this experiment is no longer active, which will
+				// cause this goroutine to be canceled. Exiting here is not helpful since the outer
+				// loop can race and restart this goroutine.
+				continue
+			}
+			for _, c := range m.criterias {
+				terminationReason, err := c.ShouldTerminate(e, dany)
```

Comment:
I know that we have a gave for the number of currently monitored experiments but could we add a counter for successful and failed (with error returned) `ShouldTerminate` calls?